### PR TITLE
Converting provinces, another part.

### PR DESCRIPTION
**Pull request by Zemurin**

### CODE REVIEW

**Comment by Zemurin** [comment]
On pause until I rework the entire converter.

File: CK3toEU4/Source/CK3World/Dynasties/Dynasties.cpp [hunk 1]
```diff
@@ -20,8 +20,13 @@ void CK3::Dynasties::registerKeys()
 		dynasties = Dynasties(theStream).getDynasties();
 	});
 	registerRegex(R"(\d+)", [this](const std::string& gameID, std::istream& theStream) {
-		const auto newDynasty = std::make_shared<Dynasty>(theStream, std::stoi(gameID));
-		dynasties.insert(std::pair(newDynasty->getGameID(), newDynasty));
+		const auto suspiciousItem = commonItems::singleItem(gameID, theStream);
+		if (suspiciousItem.find('{') != std::string::npos)
```
Comment (Idhrendur) [inline CK3toEU4/Source/CK3World/Dynasties/Dynasties.cpp:24]: You should add a test for whatever nonsense you're correcting for here.
It should fail on whatever the code was before and succeed with this change.

File: CK3toEU4/Source/CK3World/Dynasties/Houses.cpp [hunk 2]
```diff
@@ -14,8 +14,13 @@ CK3::Houses::Houses(std::istream& theStream)
 void CK3::Houses::registerKeys()
 {
 	registerRegex(R"(\d+)", [this](const std::string& ID, std::istream& theStream) {
-		const auto newHouse = std::make_shared<House>(theStream, std::stoi(ID));
-		houses.insert(std::pair(newHouse->getID(), newHouse));
+		const auto suspiciousItem = commonItems::singleItem(ID, theStream);
```
Comment (Idhrendur) [inline CK3toEU4/Source/CK3World/Dynasties/Houses.cpp:17]: Likewise, add a test for whatever input required this.

File: CK3toEU4/Source/CK3World/Geography/ProvinceHolding.cpp [hunk 3]
```diff
@@ -37,3 +37,11 @@ void CK3::ProvinceHolding::registerKeys()
 	});
 	registerRegex(commonItems::catchallRegex, commonItems::ignoreItem);
 }
+
+int CK3::ProvinceHolding::countBuildings() const
```
Comment (Idhrendur) [inline CK3toEU4/Source/CK3World/Geography/ProvinceHolding.cpp:41]: Tests

File: CK3toEU4/Source/CK3World/Titles/Title.h [hunk 4]
```diff
@@ -40,6 +44,8 @@ class Title: commonItems::parser
 	[[nodiscard]] auto isInHRE() const { return inHRE; }
 	[[nodiscard]] auto isThePope() const { return thePope; }
 	[[nodiscard]] auto isElectorate() const { return electorate; }
+	[[nodiscard]] auto isHolderCapital() const { return holderCapital; }
```
Comment (Idhrendur) [inline CK3toEU4/Source/CK3World/Titles/Title.h:47]: test

File: CK3toEU4/Source/CK3World/Titles/Title.h [hunk 5]
```diff
@@ -40,6 +44,8 @@ class Title: commonItems::parser
...
+	[[nodiscard]] auto isHRECapital() const { return HRECapital; }
```
Comment (Idhrendur) [inline CK3toEU4/Source/CK3World/Titles/Title.h:48]: test

File: CK3toEU4/Source/CK3World/Titles/Title.cpp [hunk 6]
```diff
@@ -83,6 +85,24 @@ void CK3::Title::registerKeys()
 	registerKeyword("color", [this](const std::string& unused, std::istream& theStream) {
 		color = laFabricaDeColor.getColor(theStream);
 	});
+	registerKeyword("history", [this](const std::string& unused, std::istream& theStream) {
```
Comment (Idhrendur) [inline CK3toEU4/Source/CK3World/Titles/Title.cpp:88]: test

File: CK3toEU4/Source/CK3World/Titles/Title.cpp [hunk 7]
```diff
@@ -83,6 +85,24 @@ void CK3::Title::registerKeys()
...
+		previousHolders = Title(theStream, 0).getPreviousHolders();
+	});
+	registerRegex(R"(\d+.\d+.\d+)", [this](const std::string& unused, std::istream& theStream) {
```
Comment (Idhrendur) [inline CK3toEU4/Source/CK3World/Titles/Title.cpp:91]: test(s)

File: CK3toEU4/Source/CK3World/Titles/Title.h [hunk 8]
```diff
@@ -101,33 +108,37 @@ class Title: commonItems::parser
 	void loadHoldingTitle(const std::pair<std::string, std::shared_ptr<Title>>& theTitle) { holdingTitle = theTitle; }
 	void setElectorate() { electorate = true; }
 
-	//conversion
+	// conversion
 	void loadEU4Tag(const std::pair<std::string, std::shared_ptr<EU4::Country>>& theCountry) { tagCountry = theCountry; }
-	
+	[[nodiscard]] double getBuildingWeight(const mappers::DevWeightsMapper& devWeightsMapper) const;
```
Comment (Idhrendur) [inline CK3toEU4/Source/CK3World/Titles/Title.h:113]: tests

File: CK3toEU4/Source/CK3World/Titles/Titles.cpp [hunk 9]
```diff
@@ -28,19 +28,27 @@ void CK3::Titles::registerKeys()
...
+					titles.insert(std::pair(newTitle->getName(), newTitle));
+				if (newTitle->getName().find("b_") == 0)
+					++titleCounter[0];
+				else if (newTitle->getName().find("c_") == 0)
+					++titleCounter[1];
+				else if (newTitle->getName().find("d_") == 0)
+					++titleCounter[2];
+				else if (newTitle->getName().find("k_") == 0)
+					++titleCounter[3];
+				else if (newTitle->getName().find("e_") == 0)
```
Comment (Idhrendur) [inline CK3toEU4/Source/CK3World/Titles/Titles.cpp:44]: There should be a final else case that logs an error or something.

File: CK3toEU4/Source/CK3World/Titles/Titles.cpp [hunk 10]
```diff
@@ -28,19 +28,27 @@ void CK3::Titles::registerKeys()
...
+					++titleCounter[4];
+			}
+			catch (std::exception& e)
```
Comment (Idhrendur) [inline CK3toEU4/Source/CK3World/Titles/Titles.cpp:47]: Add a test for this

File: CK3toEU4/Source/CK3World/World.cpp [hunk 11]
```diff
@@ -268,6 +268,10 @@ void CK3::World::processIronManSave(const std::string& saveGamePath)
 	auto skipLine = saveGame.gamestate.find_first_of("\r\n");
 	auto endFile = saveGame.gamestate.size();
 	saveGame.gamestate = saveGame.gamestate.substr(skipLine, endFile - skipLine);
+	// TODO: Leaving this debug in until all kinks are sorted.
```
Comment (Idhrendur) [inline CK3toEU4/Source/CK3World/World.cpp:271]: You should tie this to an issue:
```
// TODO(#42): Leaving this debug in until all kinks are sorted.
```

File: CK3toEU4/Source/EU4World/Country/Country.cpp [hunk 12]
```diff
@@ -1,6 +1,9 @@
 #include "Country.h"
```
Comment (Idhrendur) [inline CK3toEU4/Source/EU4World/Country/Country.cpp:1]: In the long term there should be tests for all of this. The java builder pattern (not to be confused with the builder pattern. Naming is hard.) I've been using in Vic2ToHoI4 can be used to set up CK3 and mapper objects for test setup.

File: CK3toEU4/Source/EU4World/Country/Country.cpp [hunk 13]
```diff
@@ -279,3 +285,909 @@ void EU4::Country::populateMisc()
...
+		details.monarch.religion = religion;
+	if (details.queen.isSet && details.queen.religion.empty())
+		details.queen.religion = religion;
+	if (details.heir.isSet && details.heir.religion.empty())
+		details.heir.religion = religion;
+}
+
+void EU4::Country::assignReforms(const std::shared_ptr<mappers::RegionMapper>& regionMapper)
+{
+	// TODO; This entire function needs love. I'm not dropping the code but many aspects won't work.
```
Comment (Idhrendur) [inline CK3toEU4/Source/EU4World/Country/Country.cpp:744]: As before, tie this to an actual issue

File: CK3toEU4/Source/EU4World/Output/outCountry.h [hunk 14]
```diff
@@ -0,0 +1,13 @@
...
+#define OUT_COUNTRY_H
+
+#include "../Country/Country.h"
+#include "../Country/CountryDetails.h"
+#include <ostream>
+
+namespace EU4
+{
+std::ostream& operator<<(std::ostream& output, const Country& country);
+std::ostream& operator<<(std::ostream& output, const Character& character);
```
Comment (Idhrendur) [inline CK3toEU4/Source/EU4World/Output/outCountry.h:11]: Wouldn't this fit better in its own file?

File: CK3toEU4/Source/Mappers/ProvinceMapper/ProvinceMapper.cpp [hunk 15]
```diff
@@ -86,12 +85,12 @@ std::map<std::string, std::shared_ptr<CK3::Title>> mappers::ProvinceMapper::getC
 					 "Province mapper error - requested ck3 province ID " + std::to_string(ck3ID) + " is invalid. We cannot find related county. Farewell.");
 			}
 		}
-		return toReturn;
 	}
 	else
 	{
-		throw std::runtime_error("Province mapper error - requested EU4 province ID " + std::to_string(eu4ProvinceNumber) + " does not exist in mappings.");
+		// We must not show fear here. Parent is likely iterating over all EU4 provinces. Play dead, they will go away.
```
Comment (Idhrendur) [inline CK3toEU4/Source/Mappers/ProvinceMapper/ProvinceMapper.cpp:91]: I must not fear. Fear is the mind-killer. Fear is the little death that brings total obliteration. I will face my fear. I will let it pass over me and through me. And when it has passed, I will turn my eye inward and see that only I remain.

Comment (Idhrendur) [reply CK3toEU4/Source/Mappers/ProvinceMapper/ProvinceMapper.cpp:91]: More practically, you can just drop the else block entirely, unless you think that will be confusing when you come back in six months or so.

**Review by Idhrendur — APPROVED** [review]

File: CK3toEU4/Source/Mappers/ProvinceMapper/ProvinceMapper.cpp (hunk 15 above)
Comment (Zemurin) [reply CK3toEU4/Source/Mappers/ProvinceMapper/ProvinceMapper.cpp:91]: That's why I leave these blocks around - to make sure I recall considering them and deciding on no action.

**Pull request merged**